uber/models.py

```diff
@@ -1625,6 +1817,464 @@ def must_contact(self):
+            consisting entirely of whitespace.
+        discount (int): The discount amount that should be applied to the
+            purchase price of a badge. The interpretation of this value
+            depends on the value of `discount_type`. In any case, a value of
+            0 equates to a full discount, i.e. a free badge.
+        discount_display (str): A human readable description of the discount.
+        discount_type (int): The type of discount this promo code will apply.
+            Valid values are:
+
+            * 0 `FIXED_DISCOUNT`: `discount` is interpreted as a fixed
+                dollar amount by which the badge price should be reduced. If
+                `discount` is 49 and the badge price is normally $100, then
+                the discounted badge price would be $51.
+
+            * 1 `FIXED_PRICE`: `discount` is interpreted as the actual badge
+                price. If `discount` is 49, then the discounted badge price
+                would be $49.
+
+            * 2 `PERCENT_DISCOUNT`: `discount` is interpreted as a percentage
+                by which the badge price should be reduced. If `discount` is
+                20 and the badge price is normally $50, then the discounted
+                badge price would $40 ($50 reduced by 20%). If `discount` is
+                100, then the price would be 100% off, i.e. a free badge.
+
+        expiration_date (datetime): The date & time upon which this promo code
+            expires. An expired promo code may no longer be used to receive
+            discounted badges.
+        is_expired (bool): True if this promo code is expired, False otherwise.
+        is_unlimited (bool): True if this promo code may be used an unlimited
+            number of times, False otherwise.
+        is_valid (bool): True if this promo code is still valid and may be
+            used again, False otherwise.
+        normalized_code (str): A normalized version of `code` suitable for
+            database queries. Normalization converts `code` to all lowercase
+            and removes dashes ("-").
+        used_by (list): List of attendees that have used this promo code.
+            Note:
+                This property is declared as a backref in the Attendee class.
+        uses_allowed (int): The total number of times this promo code may be
+            used. A value of None means this promo code may be used an
+            unlimited number of times.
+        uses_allowed_display (int): A human readable description of
+            uses_allowed.
+        uses_count (int): The number of times this promo code has already
+            been used.
+        uses_count_display (int): A human readable description of uses_count.
+        uses_remaining (int): Remaining number of times this promo code may
+            be used.
+        uses_remaining_display (int): A human readable description of
+            uses_remaining.
+    """
+
+    FIXED_DISCOUNT = 0
```

Comment:
You raise an interesting point, and I generally agree that configuration variables should be stored in the config. I think the debate is arising because `DISCOUNT_OPTS` LOOKS like a config variable (indeed it is named and structured in a similar fashion), but `DISCOUNT_OPTS` is definitely NOT a config variable.

`DISCOUNT_OPTS` is intrinsic to the operation of `PromoCode`. The `PromoCode` class understands each discount type, and has well defined behavior associated with each value. Consider what would happen if we stored these values in our config files, and someone altered them. The `PromoCode` class would immediately break because it relies on those values existing.

If we wanted to store something in the configuration, it would be a list of enabled discount types, which we would use to filter `DISCOUNT_OPTS`. For example, we could configure a server to only allow percent discounts, or only allow fixed price discounts. But the `DISCOUNT_OPTS` themselves would still live in the model class.

However – and this is a larger point that I want to emphasize – even if we wanted to give administrators the ability to disallow certain discount types we should NOT put that in the config. That should be an additional column in the database, like `enabled_discount_opts` or something.

Anything we put in the config requires a programmer to modify in GitHub, and a server redeploy to make the changes in production. There's no reason why a programmer should be involved if an event leader wants to change some event configuration. Want to change registration price bumps? Here's a nice form were you can modify those values.